File: service/src/test/java/no/nav/melosys/service/medl/MedlPeriodeServiceTest.java
package no.nav.melosys.service.medl;

import java.time.LocalDate;
import java.util.List;
import java.util.Set;

import no.nav.melosys.domain.*;
import no.nav.melosys.domain.kodeverk.Aktoersroller;
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingsstatus;
import no.nav.melosys.exception.FunksjonellException;
import no.nav.melosys.integrasjon.medl.KildedokumenttypeMedl;
import no.nav.melosys.integrasjon.medl.MedlService;
import no.nav.melosys.integrasjon.medl.StatusaarsakMedl;
import no.nav.melosys.repository.AnmodningsperiodeRepository;
import no.nav.melosys.repository.LovvalgsperiodeRepository;
import no.nav.melosys.repository.MedlemskapsperiodeRepository;
import no.nav.melosys.repository.UtpekingsperiodeRepository;
import no.nav.melosys.service.behandling.BehandlingsresultatService;
import no.nav.melosys.service.persondata.PersondataFasade;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class MedlPeriodeServiceTest {

    @Mock
    private PersondataFasade persondataFasade;
    @Mock
    private MedlService medlService;
    @Mock
    private BehandlingsresultatService behandlingsresultatService;
    @Mock
    private LovvalgsperiodeRepository lovvalgsperiodeRepository;
    @Mock
    private AnmodningsperiodeRepository anmodningsperiodeRepository;
    @Mock
    private UtpekingsperiodeRepository utpekingsperiodeRepository;
    @Mock
    private MedlemskapsperiodeRepository medlemskapsperiodeRepository;

    private MedlPeriodeService medlPeriodeService;

    private static final String FNR = "12345678901";
    private static final Long MEDL_PERIODE_ID = 99L;

    @BeforeEach
    public void setUp() throws Exception {
        medlPeriodeService = new MedlPeriodeService(persondataFasade, medlService, behandlingsresultatService,
            lovvalgsperiodeRepository, anmodningsperiodeRepository, utpekingsperiodeRepository, medlemskapsperiodeRepository);
    }

    @Test
    void hentPeriodeListe() {
        medlPeriodeService.hentPeriodeListe(FNR, LocalDate.now(), LocalDate.now().plusMonths(2));

        verify(medlService).hentPeriodeListe(eq(FNR), any(LocalDate.class), any(LocalDate.class));
    }

    @Test
    void opprettPeriodeForeløpig() {
        setupPeriodeForeløpig();

        medlPeriodeService.opprettPeriodeForeløpig(new Utpekingsperiode(), 1L, true);

        verify(medlService).opprettPeriodeForeløpig(eq(FNR), any(Utpekingsperiode.class), eq(KildedokumenttypeMedl.SED));
        verify(utpekingsperiodeRepository).save(any(Utpekingsperiode.class));
    }

    @Test
    void opprettPeriodeUnderAvklaring() {
        setupPeriodeUnderAvklaring();

        medlPeriodeService.opprettPeriodeUnderAvklaring(new Anmodningsperiode(), 1L, false);

        verify(medlService).opprettPeriodeUnderAvklaring(eq(FNR), any(Anmodningsperiode.class), eq(KildedokumenttypeMedl.HENV_SOKNAD));
        verify(anmodningsperiodeRepository).save(any(Anmodningsperiode.class));
    }

    @Test
    void opprettPeriodeEndelig() {
        setupPeriodeEndelig();

        medlPeriodeService.opprettPeriodeEndelig(new Lovvalgsperiode(), 1L, true);

        verify(medlService).opprettPeriodeEndelig(eq(FNR), any(Lovvalgsperiode.class), eq(KildedokumenttypeMedl.SED));
        verify(lovvalgsperiodeRepository).save(any(Lovvalgsperiode.class));
    }

    @Test
    void opprettPeriodeEndeligFtrl_feiler() throws Exception {
        setupHappyPathBehandling();
        when(medlService.opprettPeriodeEndelig(eq(FNR), any(Medlemskapsperiode.class), any(KildedokumenttypeMedl.class)))
            .thenReturn(null);

        assertThatThrownBy(() -> medlPeriodeService.opprettPeriodeEndelig(1L, new Medlemskapsperiode()))
            .isInstanceOf(FunksjonellException.class)
            .hasMessageContaining("Opprettelse av periode i MEDL feilet med retur av null medlPeriodeID");

        verifyNoInteractions(medlemskapsperiodeRepository);
    }

    @Test
    void opprettPeriodeEndeligFtrl() throws Exception {
        setupHappyPathBehandling();

        medlPeriodeService.opprettPeriodeEndelig(1L, new Medlemskapsperiode());

        verify(medlService).opprettPeriodeEndelig(eq(FNR), any(Medlemskapsperiode.class), any(KildedokumenttypeMedl.class));
        verify(medlemskapsperiodeRepository).save(any(Medlemskapsperiode.class));
    }

    @Test
    void oppdaterPeriodeEndelig() {
        Lovvalgsperiode lovvalgsperiode = new Lovvalgsperiode();
        lovvalgsperiode.setMedlPeriodeID(MEDL_PERIODE_ID);
        medlPeriodeService.oppdaterPeriodeEndelig(lovvalgsperiode, false);

        verify(medlService).oppdaterPeriodeEndelig(lovvalgsperiode, KildedokumenttypeMedl.HENV_SOKNAD);
    }

    @Test
    void oppdaterPeriodeForeløpig() {
        Lovvalgsperiode lovvalgsperiode = new Lovvalgsperiode();
        lovvalgsperiode.setMedlPeriodeID(MEDL_PERIODE_ID);
        medlPeriodeService.oppdaterPeriodeForeløpig(lovvalgsperiode, false);

        verify(medlService).oppdaterPeriodeForeløpig(lovvalgsperiode, KildedokumenttypeMedl.HENV_SOKNAD);
    }

    @Test
    void avvisPeriode() {
        medlPeriodeService.avvisPeriode(MEDL_PERIODE_ID);
        verify(medlService).avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.AVVIST);
    }

    @Test
    void avvisPeriodeFeilregistrert() {
        medlPeriodeService.avvisPeriodeFeilregistrert(MEDL_PERIODE_ID);
        verify(medlService).avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.FEILREGISTRERT);
    }

    @Test
    void avvisPeriodeOpphørt() {
        medlPeriodeService.avvisPeriodeOpphørt(MEDL_PERIODE_ID);
        verify(medlService).avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.OPPHORT);
    }

    @Test
    void avsluttTidligereMedlPeriode_behandlingOgPeriodeFinnes_avviserPeriode() {
        Behandling behandling = new Behandling();
        behandling.setStatus(Behandlingsstatus.AVSLUTTET);
        behandling.setId(1L);

        Fagsak fagsak = new Fagsak();
        fagsak.setBehandlinger(List.of(behandling));

        Lovvalgsperiode lovvalgsperiode = new Lovvalgsperiode();
        lovvalgsperiode.setMedlPeriodeID(MEDL_PERIODE_ID);
        Behandlingsresultat behandlingsresultat = new Behandlingsresultat();
        behandlingsresultat.setLovvalgsperioder(Set.of(lovvalgsperiode));
        when(behandlingsresultatService.hentBehandlingsresultat(1L)).thenReturn(behandlingsresultat);

        medlPeriodeService.avsluttTidligerMedlPeriode(fagsak);

        verify(behandlingsresultatService).hentBehandlingsresultat(1L);
        verify(medlService).avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.AVVIST);
    }

    @Test
    void avsluttTidligereMedlPeriode_ingenEksisterendePeriode_ingenPeriodeBlirAvvist() {
        Behandling behandling = new Behandling();
        behandling.setStatus(Behandlingsstatus.AVSLUTTET);
        behandling.setId(1L);

        Fagsak fagsak = new Fagsak();
        fagsak.setBehandlinger(List.of(behandling));

        Behandlingsresultat behandlingsresultat = new Behandlingsresultat();
        behandlingsresultat.setLovvalgsperioder(Set.of());
        when(behandlingsresultatService.hentBehandlingsresultat(1L)).thenReturn(behandlingsresultat);

        medlPeriodeService.avsluttTidligerMedlPeriode(fagsak);

        verify(behandlingsresultatService).hentBehandlingsresultat(1L);
        verify(medlService, never()).avvisPeriode(anyLong(), any(StatusaarsakMedl.class));
    }

    private void setupPeriodeForeløpig() {
        when(medlService.opprettPeriodeForeløpig(anyString(), any(PeriodeOmLovvalg.class), any(KildedokumenttypeMedl.class)))
            .thenReturn(MEDL_PERIODE_ID);
        setupHappyPathBehandling();
    }

    private void setupPeriodeUnderAvklaring() {
        when(medlService.opprettPeriodeUnderAvklaring(anyString(), any(PeriodeOmLovvalg.class), any(KildedokumenttypeMedl.class)))
            .thenReturn(MEDL_PERIODE_ID);
        setupHappyPathBehandling();
    }

    private void setupPeriodeEndelig() {
        when(medlService.opprettPeriodeEndelig(anyString(), any(Lovvalgsperiode.class), any(KildedokumenttypeMedl.class)))
            .thenReturn(MEDL_PERIODE_ID);
        setupHappyPathBehandling();
    }

    private void setupHappyPathBehandling() {
        when(behandlingsresultatService.hentBehandlingsresultat(anyLong())).thenReturn(lagBehandlingsResultat());
        when(persondataFasade.hentFolkeregisterIdent(anyString())).thenReturn(FNR);
    }

    private Behandlingsresultat lagBehandlingsResultat() {
        Behandlingsresultat behandlingsresultat = new Behandlingsresultat();
        Behandling behandling = new Behandling();
        Fagsak fagsak = new Fagsak();

        Aktoer aktoer = new Aktoer();
        aktoer.setRolle(Aktoersroller.BRUKER);
        aktoer.setAktørId("456");
        fagsak.getAktører().add(aktoer);

        behandling.setFagsak(fagsak);
        behandlingsresultat.setBehandling(behandling);
        return behandlingsresultat;
    }
}

File: service/src/test/kotlin/no/nav/melosys/service/medl/MedlPeriodeServiceTest.kt
package no.nav.melosys.service.medl

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.string.shouldContain
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import no.nav.melosys.domain.*
import no.nav.melosys.domain.kodeverk.Aktoersroller
import no.nav.melosys.domain.kodeverk.Sakstyper
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingsstatus
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstema
import no.nav.melosys.exception.TekniskException
import no.nav.melosys.integrasjon.medl.KildedokumenttypeMedl
import no.nav.melosys.integrasjon.medl.MedlService
import no.nav.melosys.integrasjon.medl.StatusaarsakMedl
import no.nav.melosys.repository.AnmodningsperiodeRepository
import no.nav.melosys.repository.LovvalgsperiodeRepository
import no.nav.melosys.repository.MedlemskapsperiodeRepository
import no.nav.melosys.repository.UtpekingsperiodeRepository
import no.nav.melosys.service.behandling.BehandlingService
import no.nav.melosys.service.behandling.BehandlingsresultatService
import no.nav.melosys.service.persondata.PersondataFasade
import no.nav.melosys.service.sak.FagsakService
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.time.LocalDate


@ExtendWith(MockKExtension::class)
class MedlPeriodeServiceTest {
    @MockK
    lateinit var persondataFasade: PersondataFasade

    @RelaxedMockK
    lateinit var medlService: MedlService

    @MockK
    lateinit var behandlingsresultatService: BehandlingsresultatService

    @MockK
    lateinit var behandlingService: BehandlingService

    @MockK
    lateinit var lovvalgsperiodeRepository: LovvalgsperiodeRepository

    @MockK
    lateinit var anmodningsperiodeRepository: AnmodningsperiodeRepository

    @RelaxedMockK
    lateinit var medlAnmodningsperiodeService: MedlAnmodningsperiodeService

    @MockK
    lateinit var utpekingsperiodeRepository: UtpekingsperiodeRepository

    @MockK
    lateinit var medlemskapsperiodeRepository: MedlemskapsperiodeRepository

    @MockK
    lateinit var fagsakService: FagsakService

    lateinit var medlPeriodeService: MedlPeriodeService


    @BeforeEach
    fun setUp() {
        medlPeriodeService = MedlPeriodeService(
            persondataFasade,
            medlService,
            behandlingsresultatService,
            behandlingService,
            lovvalgsperiodeRepository,
            medlAnmodningsperiodeService,
            utpekingsperiodeRepository,
            medlemskapsperiodeRepository,
            fagsakService
        )
    }

    @Test
    fun hentPeriodeListe() {
        val now = LocalDate.now()
        val plusMonths = LocalDate.now().plusMonths(2)

        medlPeriodeService.hentPeriodeListe(FNR, now, plusMonths)

        verify { medlService.hentPeriodeListe(FNR, now, plusMonths) }
    }

    @Test
    fun opprettPeriodeForeløpig() {
        setupHappyPathBehandling()
        val utpekingsperiode = Utpekingsperiode()
        every { medlService.opprettPeriodeForeløpig(any(), any(), any()) } returns MEDL_PERIODE_ID
        every { utpekingsperiodeRepository.save(any()) } returns utpekingsperiode

        medlPeriodeService.opprettPeriodeForeløpig(utpekingsperiode, 1)

        verify { medlService.opprettPeriodeForeløpig(FNR, any(), KildedokumenttypeMedl.SED) }
        verify { utpekingsperiodeRepository.save(utpekingsperiode) }
    }

    @Test
    fun opprettPeriodeUnderAvklaring() {
        setupHappyPathBehandling(Sakstyper.EU_EOS, Behandlingstema.UTSENDT_ARBEIDSTAKER)
        every { medlService.opprettPeriodeUnderAvklaring(any(), any(), any()) } returns MEDL_PERIODE_ID
        every { anmodningsperiodeRepository.save(any()) } returns Anmodningsperiode()

        medlPeriodeService.opprettPeriodeUnderAvklaring(Anmodningsperiode(), 1L)
        verify { medlService.opprettPeriodeUnderAvklaring(FNR, any(), KildedokumenttypeMedl.HENV_SOKNAD) }
        verify { medlAnmodningsperiodeService.lagreAnmodningsperiode(any()) }
    }

    @Test
    fun opprettPeriodeEndelig() {
        setupHappyPathBehandling()
        every { medlService.opprettPeriodeEndelig(any(), any(), any()) } returns MEDL_PERIODE_ID
        every { lovvalgsperiodeRepository.save(any()) } returns Lovvalgsperiode()

        medlPeriodeService.opprettPeriodeEndelig(Lovvalgsperiode(), 1L)

        verify { medlService.opprettPeriodeEndelig(FNR, any(), KildedokumenttypeMedl.SED) }
        verify { lovvalgsperiodeRepository.save(any()) }
    }

    @Test
    fun opprettPeriodeEndeligFtrl_feiler() {
        setupHappyPathBehandling()
        every { medlService.opprettPeriodeEndelig(FNR, any(), any()) } returns null

        shouldThrow<TekniskException> {
            medlPeriodeService.opprettPeriodeEndelig(1L, Medlemskapsperiode())
        }.message.shouldContain("Opprettelse av periode i MEDL feilet med retur av null medlPeriodeID")

        verify(exactly = 0) { medlemskapsperiodeRepository.save(any()) }
    }

    @Test
    fun opprettPeriodeEndeligFtrl() {
        setupHappyPathBehandling()
        val medlemskapsperiode = Medlemskapsperiode()
        every { medlemskapsperiodeRepository.save(any()) } returns medlemskapsperiode

        medlPeriodeService.opprettPeriodeEndelig(1L, medlemskapsperiode)

        verify { medlService.opprettPeriodeEndelig(FNR, any(), any()) }
        verify { medlemskapsperiodeRepository.save(medlemskapsperiode) }
    }

    @Test
    fun opprettOpphørtPeriode() {
        setupHappyPathBehandling()
        val medlemskapsperiode = Medlemskapsperiode()
        every { medlemskapsperiodeRepository.save(any()) } returns medlemskapsperiode

        medlPeriodeService.opprettOpphørtPeriode(1L, medlemskapsperiode)

        verify { medlService.opprettOpphørtPeriode(FNR, any(), any()) }
        verify { medlemskapsperiodeRepository.save(medlemskapsperiode) }
    }


    @Test
    fun oppdaterPeriodeEndelig() {
        setupHappyPathBehandling(Sakstyper.EU_EOS, Behandlingstema.UTSENDT_ARBEIDSTAKER)

        val lovvalgsperiode = Lovvalgsperiode().apply {
            medlPeriodeID = MEDL_PERIODE_ID
            behandlingsresultat = lagBehandlingsResultat()
        }


        medlPeriodeService.oppdaterPeriodeEndelig(lovvalgsperiode)


        verify { medlService.oppdaterPeriodeEndelig(lovvalgsperiode, KildedokumenttypeMedl.HENV_SOKNAD) }
    }

    @Test
    fun oppdaterOpphørtPeriode() {
        setupHappyPathBehandling(Sakstyper.FTRL, Behandlingstema.YRKESAKTIV)
        val medlemskapsperiode = Medlemskapsperiode().apply {
            medlPeriodeID = MEDL_PERIODE_ID
        }


        medlPeriodeService.oppdaterOpphørtPeriode(1, medlemskapsperiode)


        verify { medlService.oppdaterOpphørtPeriode(medlemskapsperiode, KildedokumenttypeMedl.HENV_SOKNAD) }
    }

    @Test
    fun oppdaterPeriodeForeløpig() {
        setupHappyPathBehandling(Sakstyper.EU_EOS, Behandlingstema.UTSENDT_ARBEIDSTAKER)
        val lovvalgsperiode = Lovvalgsperiode().apply {
            medlPeriodeID = MEDL_PERIODE_ID
            behandlingsresultat = lagBehandlingsResultat()
        }


        medlPeriodeService.oppdaterPeriodeForeløpig(lovvalgsperiode)


        verify { medlService.oppdaterPeriodeForeløpig(lovvalgsperiode, KildedokumenttypeMedl.HENV_SOKNAD) }
    }

    @Test
    fun `oppdaterPeriodeForeløpig bruker KildedokumenttypeMedl DOKUMENT når TRYGDEAVTALE og REGISTRERING_UNNTAK`() {
        every { behandlingService.hentBehandling(any()) } returns Behandling().apply {
            tema = Behandlingstema.REGISTRERING_UNNTAK
            fagsak = FagsakTestFactory.builder().type(Sakstyper.TRYGDEAVTALE).medBruker().build()
        }
        val lovvalgsperiode = Lovvalgsperiode().apply {
            medlPeriodeID = MEDL_PERIODE_ID
            behandlingsresultat = lagBehandlingsResultat()
        }


        medlPeriodeService.oppdaterPeriodeForeløpig(lovvalgsperiode)


        verify { medlService.oppdaterPeriodeForeløpig(lovvalgsperiode, KildedokumenttypeMedl.DOKUMENT) }
    }

    @Test
    fun `oppdaterPeriodeForeløpig bruker KildedokumenttypeMedl HENV_SOKNAD når TRYGDEAVTALE og ANMODNING_OM_UNNTAK`() {
        every { behandlingService.hentBehandling(any()) } returns Behandling().apply {
            tema = Behandlingstema.ANMODNING_OM_UNNTAK_HOVEDREGEL
            fagsak = FagsakTestFactory.builder().type(Sakstyper.TRYGDEAVTALE).medBruker().build()
        }

        val lovvalgsperiode = Lovvalgsperiode().apply {
            medlPeriodeID = MEDL_PERIODE_ID
            behandlingsresultat = lagBehandlingsResultat()
        }


        medlPeriodeService.oppdaterPeriodeForeløpig(lovvalgsperiode)


        verify { medlService.oppdaterPeriodeForeløpig(lovvalgsperiode, KildedokumenttypeMedl.HENV_SOKNAD) }
    }

    @Test
    fun `oppdaterPeriodeForeløpig bruker KildedokumenttypeMedl A1 når EU_EOS og A1_ANMODNING_OM_UNNTAK_PAPIR`() {
        every { behandlingService.hentBehandling(any()) } returns Behandling().apply {
            tema = Behandlingstema.A1_ANMODNING_OM_UNNTAK_PAPIR
            fagsak = FagsakTestFactory.builder().medBruker().build()
        }
        val lovvalgsperiode = Lovvalgsperiode().apply {
            medlPeriodeID = MEDL_PERIODE_ID
            behandlingsresultat = lagBehandlingsResultat()
        }


        medlPeriodeService.oppdaterPeriodeForeløpig(lovvalgsperiode)


        verify { medlService.oppdaterPeriodeForeløpig(lovvalgsperiode, KildedokumenttypeMedl.A1) }
    }

    @Test
    fun avvisPeriode() {
        medlPeriodeService.avvisPeriode(MEDL_PERIODE_ID)

        verify { medlService.avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.AVVIST) }
    }

    @Test
    fun avvisPeriodeFeilregistrert() {
        medlPeriodeService.avvisPeriodeFeilregistrert(MEDL_PERIODE_ID)

        verify { medlService.avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.FEILREGISTRERT) }
    }

    @Test
    fun avvisPeriodeOpphørt() {
        medlPeriodeService.avvisPeriodeOpphørt(MEDL_PERIODE_ID)

        verify { medlService.avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.OPPHORT) }
    }

    @Test
    fun avsluttTidligereMedlPeriode_behandlingOgPeriodeFinnes_avviserPeriode() {
        val fagsak = FagsakTestFactory.builder()
            .behandlinger(
                Behandling().apply {
                    status = Behandlingsstatus.AVSLUTTET
                    id = 1L
                }
            ).build()
        val behandlingsresultat = Behandlingsresultat().apply {
            lovvalgsperioder =
                setOf(Lovvalgsperiode().apply { medlPeriodeID = MEDL_PERIODE_ID })
        }
        every { behandlingsresultatService.hentBehandlingsresultat(1L) } returns behandlingsresultat
        every { fagsakService.hentFagsak("MEL-1") } returns fagsak

        medlPeriodeService.avsluttTidligerMedlPeriode("MEL-1")

        verify { behandlingsresultatService.hentBehandlingsresultat(1L) }
        verify { medlService.avvisPeriode(MEDL_PERIODE_ID, StatusaarsakMedl.AVVIST) }
    }

    @Test
    fun avsluttTidligereMedlPeriode_ingenEksisterendePeriode_ingenPeriodeBlirAvvist() {
        val fagsak = FagsakTestFactory.builder()
            .behandlinger(
                Behandling().apply {
                    status = Behandlingsstatus.AVSLUTTET
                    id = 1L
                }
            ).build()
        val behandlingsresultat = Behandlingsresultat().apply { lovvalgsperioder = setOf() }
        every { behandlingsresultatService.hentBehandlingsresultat(1L) } returns behandlingsresultat
        every { fagsakService.hentFagsak("MEL-1") } returns fagsak

        medlPeriodeService.avsluttTidligerMedlPeriode("MEL-1")

        verify { behandlingsresultatService.hentBehandlingsresultat(1L) }
        verify(exactly = 0) { medlService.avvisPeriode(any(), any()) }
    }

    private fun setupHappyPathBehandling(
        sakstype: Sakstyper = Sakstyper.EU_EOS,
        behandlingstema: Behandlingstema = Behandlingstema.BESLUTNING_LOVVALG_NORGE
    ) {
        val behandlingResultat = lagBehandlingsResultat(sakstype, behandlingstema)
        every { behandlingsresultatService.hentBehandlingsresultat(any()) } returns behandlingResultat
        every { persondataFasade.hentFolkeregisterident(any()) } returns FNR
        every { behandlingService.hentBehandling(any()) } returns behandlingResultat.behandling
    }

    private fun lagBehandlingsResultat(
        sakstype: Sakstyper = Sakstyper.EU_EOS,
        behandlingstema: Behandlingstema = Behandlingstema.BESLUTNING_LOVVALG_NORGE
    ):
        Behandlingsresultat = Behandlingsresultat().apply {
        id = 1L
        behandling = Behandling().apply {
            tema = behandlingstema
            fagsak = FagsakTestFactory.builder().type(sakstype).medBruker().build()
        }
    }

    companion object {
        private const val FNR = "12345678901"
        private const val MEDL_PERIODE_ID = 99L
    }
}
